Handle "payment succeeded" stripe webhook

jurgenwerk opened the pull request:
This will allow us to listen to the `invoice.payment_succeeded` webhook from Stripe. This webhook includes `billing_reason: 'subscription_create' | 'subscription_cycle';` which we use to see whether to create a subscription or to renew it.

jurgenwerk committed:
Happy path for creating a new subscription from a stripe invoice.paid webhook

jurgenwerk committed:
Add subscription renewal flow

jurgenwerk committed:
Apply types

jurgenwerk committed:
Add tx manager

jurgenwerk committed:
Improvements to the payment succeeded handler

jurgenwerk committed:
Not needed

jurgenwerk committed:
Should use STRIPE_WEBHOOK_SECRET

jurgenwerk committed:
Merge branch 'main' into cs-7386-handle-stripe-subscribed-webhook

jurgenwerk committed:
Update after table renaming

jurgenwerk committed:
Cleanup

jurgenwerk committed:
Fix getting product id

jurgenwerk committed:
Cleanup

jurgenwerk commented on packages/realm-server/migrations/1729595943510_billing-adjustments.js:
```diff
@@ -0,0 +1,68 @@
+exports.shorthands = undefined;
+
+exports.up = (pgm) => {
+  pgm.renameTable('credit_balance_events', 'credits_ledger');
```
It's a more intuitive name

jurgenwerk commented on packages/realm-server/migrations/1729595943510_billing-adjustments.js:
```diff
@@ -0,0 +1,68 @@
+    notNull: false,
+  });
+
+  pgm.alterColumn('subscriptions', 'started_at', {
```
Changing timestamp types to unix time so that we can directly sync with Stripe's timestamps which also come as unix time

habdelra replied:
unix times do not need a `bigint`--they can fit comfortably as a normal integer

jurgenwerk commented on packages/realm-server/billing/stripe-webhook-handlers/payment-succeeded.ts:
```diff
@@ -0,0 +1,148 @@
+      event.data.object.lines.data[0].price.product,
+    );
+
+    // When user first signs up for a plan, our checkout.session.completed handler takes care of assigning the user a stripe customer id.
```
Comment is mentioning https://github.com/cardstack/boxel/pull/1723

backspace commented on packages/realm-server/package.json:
```diff
@@ -100,5 +100,8 @@
     "extends": "../../package.json"
+  },
+  "dependencies": {
+    "stripe": "^17.2.1"
```
everything else is in `devDependencies`, does this need to be here instead?

habdelra replied:
agreed--this must be an oversight. since no packages consume realm-server there is no need to ever have `dependencies` in the package.json

backspace commented on packages/realm-server/migrations/1729595943510_billing-adjustments.js:
```diff
@@ -0,0 +1,68 @@
+  });
+
+  pgm.alterColumn('subscriptions', 'ended_at', {
+    type: 'timestamp',
```
Given the below is this also meant to be `bigint`?

habdelra replied:
confirm, we should treat dates consistently, not as numbers sometimes and date objects other times

backspace commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+        let user = await insertUser(dbAdapter, 'user@test', 'cus_123');
+        let plan = await insertPlan(dbAdapter, 'Free plan', 0, 100, 'prod_123');
+
+        // Omitted version of a real stripe invoice.payment_succeeded event
```
What does `Omitted` mean here?

jurgenwerk replied:
The actual event contains many more fields, but our type defines only the ones we care about 

Here is an example of a full event:

```
{
  "object": {
    "id": "[REDACTED_INVOICE_ID]",
    "object": "invoice",
    "account_country": "US",
    "account_name": "[REDACTED_COMPANY_NAME]",
    "account_tax_ids": null,
    "amount_due": 0,
    "amount_paid": 0,
    "amount_remaining": 0,
    "amount_shipping": 0,
    "application": null,
    "application_fee_amount": null,
    "attempt_count": 0,
    "attempted": true,
    "auto_advance": false,
    "automatic_tax": {
      "enabled": true,
      "liability": {
        "type": "self"
      },
      "status": "complete"
    },
    "automatically_finalizes_at": null,
    "billing_reason": "subscription_create",
    "charge": null,
    "collection_method": "charge_automatically",
    "created": "[REDACTED_TIMESTAMP]",
    "currency": "usd",
    "custom_fields": null,
    "customer": "[REDACTED_CUSTOMER_ID]",
    "customer_address": {
      "city": null,
      "country": "[REDACTED]",
      "line1": null,
      "line2": null,
      "postal_code": null,
      "state": null
    },
    "customer_email": "[REDACTED_EMAIL]",
    "customer_name": "[REDACTED]",
    "customer_phone": null,
    "customer_shipping": null,
    "customer_tax_exempt": "none",
    "customer_tax_ids": [],
    "default_payment_method": null,
    "default_source": null,
    "default_tax_rates": [],
    "description": null,
    "discount": null,
    "discounts": [],
    "due_date": null,
    "effective_at": "[REDACTED_TIMESTAMP]",
    "ending_balance": 0,
    "footer": null,
    "from_invoice": null,
    "hosted_invoice_url": "[REDACTED_URL]",
    "invoice_pdf": "[REDACTED_URL]",
    "issuer": {
      "type": "self"
    },
    "last_finalization_error": null,
    "latest_revision": null,
    "lines": {
      "object": "list",
      "data": [{
        "id": "[REDACTED_LINE_ITEM_ID]",
        "object": "line_item",
        "amount": 0,
        "amount_excluding_tax": 0,
        "currency": "usd",
        "description": "1 × Free (at $0.00 / month)",
        "discount_amounts": [],
        "discountable": true,
        "discounts": [],
        "invoice": "[REDACTED_INVOICE_ID]",
        "livemode": false,
        "metadata": {},
        "period": {
          "end": "[REDACTED_TIMESTAMP]",
          "start": "[REDACTED_TIMESTAMP]"
        },
        "plan": {
          "id": "[REDACTED_PLAN_ID]",
          "object": "plan",
          "active": true,
          "aggregate_usage": null,
          "amount": 0,
          "amount_decimal": "0",
          "billing_scheme": "per_unit",
          "created": "[REDACTED_TIMESTAMP]",
          "currency": "usd",
          "interval": "month",
          "interval_count": 1,
          "livemode": false,
          "metadata": {},
          "meter": null,
          "nickname": null,
          "product": "[REDACTED_PRODUCT_ID]",
          "tiers_mode": null,
          "transform_usage": null,
          "trial_period_days": null,
          "usage_type": "licensed"
        },
        "pretax_credit_amounts": [],
        "price": {
          "id": "[REDACTED_PRICE_ID]",
          "object": "price",
          "active": true,
          "billing_scheme": "per_unit",
          "created": "[REDACTED_TIMESTAMP]",
          "currency": "usd",
          "custom_unit_amount": null,
          "livemode": false,
          "lookup_key": null,
          "metadata": {},
          "nickname": null,
          "product": "[REDACTED_PRODUCT_ID]",
          "recurring": {
            "aggregate_usage": null,
            "interval": "month",
            "interval_count": 1,
            "meter": null,
            "trial_period_days": null,
            "usage_type": "licensed"
          },
          "tax_behavior": "unspecified",
          "tiers_mode": null,
          "transform_quantity": null,
          "type": "recurring",
          "unit_amount": 0,
          "unit_amount_decimal": "0"
        },
        "proration": false,
        "proration_details": {
          "credited_items": null
        },
        "quantity": 1,
        "subscription": "[REDACTED_SUBSCRIPTION_ID]",
        "subscription_item": "[REDACTED_SUBSCRIPTION_ITEM_ID]",
        "tax_amounts": [{
          "amount": 0,
          "inclusive": false,
          "tax_rate": "[REDACTED_TAX_RATE_ID]",
          "taxability_reason": "not_collecting",
          "taxable_amount": 0
        }],
        "tax_rates": [],
        "type": "subscription",
        "unit_amount_excluding_tax": "0"
      }],
      "has_more": false,
      "total_count": 1,
      "url": "/v1/invoices/[REDACTED_INVOICE_ID]/lines"
    },
    "livemode": false,
    "metadata": {},
    "next_payment_attempt": null,
    "number": "[REDACTED_INVOICE_NUMBER]",
    "on_behalf_of": null,
    "paid": true,
    "paid_out_of_band": false,
    "payment_intent": null,
    "payment_settings": {
      "default_mandate": null,
      "payment_method_options": {
        "acss_debit": null,
        "bancontact": null,
        "card": {
          "request_three_d_secure": "automatic"
        },
        "customer_balance": null,
        "konbini": null,
        "sepa_debit": null,
        "us_bank_account": null
      },
      "payment_method_types": null
    },
    "period_end": "[REDACTED_TIMESTAMP]",
    "period_start": "[REDACTED_TIMESTAMP]",
    "post_payment_credit_notes_amount": 0,
    "pre_payment_credit_notes_amount": 0,
    "quote": null,
    "receipt_number": null,
    "rendering": null,
    "shipping_cost": null,
    "shipping_details": null,
    "starting_balance": 0,
    "statement_descriptor": null,
    "status": "paid",
    "status_transitions": {
      "finalized_at": "[REDACTED_TIMESTAMP]",
      "marked_uncollectible_at": null,
      "paid_at": "[REDACTED_TIMESTAMP]",
      "voided_at": null
    },
    "subscription": "[REDACTED_SUBSCRIPTION_ID]",
    "subscription_details": {
      "metadata": {}
    },
    "subtotal": 0,
    "subtotal_excluding_tax": 0,
    "tax": 0,
    "test_clock": null,
    "total": 0,
    "total_discount_amounts": [],
    "total_excluding_tax": 0,
    "total_pretax_credit_amounts": [],
    "total_tax_amounts": [{
      "amount": 0,
      "inclusive": false,
      "tax_rate": "[REDACTED_TAX_RATE_ID]",
      "taxability_reason": "not_collecting",
      "taxable_amount": 0
    }],
    "transfer_data": null,
    "webhooks_delivered_at": null
  },
  "previous_attributes": null
}
```

backspace commented on packages/realm-server/billing/stripe-webhook-handlers/index.ts:
```diff
@@ -0,0 +1,80 @@
+  let type = event.type;
+
+  // For adding extra credits, we should listen for charge.succeeded, and for
+  // subsciptions, we should listen for invoice.payment_succeeded (I discovered this when I was
```
```suggestion
  // subscriptions, we should listen for invoice.payment_succeeded (I discovered this when I was
```

backspace reviewed (approved):
This is my first time looking in depth at this feature, I mostly understood it through the tests, looks good

lukemelia commented on packages/realm-server/billing/stripe-webhook-handlers/payment-succeeded.ts:
```diff
@@ -0,0 +1,148 @@
+    await query(dbAdapter, [
+      `UPDATE stripe_events SET is_processed = TRUE WHERE stripe_event_id = `,
+      param(event.id),
+    ]);
```
This is a long method. It is pretty clear, which is great, but might benefit from being broken down into some smaller methods with expressive names.

habdelra replied:
it seems awkward that you have this one bit of SQL in this method, where everything else is invoking functions to execute SQL. perhaps this SQL should get moved into the `billing-queries.ts`

lukemelia commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+    creditType: result.credit_type as string,
+    subscriptionCycleId: result.subscription_cycle_id as string,
+  }));
+}
```
We will probably want to extract these to a test helpers file at some point. OK to be here for now though.

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    event_data: event.data,
+  });
+  await query(dbAdapter, [
+    `INSERT INTO stripe_events (stripe_event_id, event_type, event_data) VALUES`,
```
you can actually use the `nameExpressions` instead of having to enter the column names--that's the whole point of that variable

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    throw new Error(`No plan found with stripe plan id: ${stripePlanId}`);
+  }
+
+  return {
```
use the `Plan` table type that you defined above instead of having to recast each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    return null;
+  }
+
+  return {
```
ditto--use the `User` table type instead of having to recast each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+  });
+
+  let result = await query(dbAdapter, [
+    `INSERT INTO subscription_cycles (subscription_id, period_start, period_end) VALUES`,
```
use the `nameExpression` instead of having to enter each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    ` RETURNING *`,
+  ] as Expression);
+
+  return {
```
use the `SubscriptionCycle` type instead of having to cast each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+  });
+
+  let result = await query(dbAdapter, [
+    `INSERT INTO subscriptions (user_id, plan_id, started_at, status, stripe_subscription_id) VALUES`,
```
use the `nameExpression`  instead of having to type each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    ` RETURNING *`,
+  ] as Expression);
+
+  return {
```
use the `Subscription` type instead of casting each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+  });
+
+  await query(dbAdapter, [
+    `INSERT INTO credits_ledger (user_id, credit_amount, credit_type, subscription_cycle_id) VALUES`,
```
use the `nameExpression` instead of having to enter each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    return null;
+  }
+
+  return results[0];
```
use the `StripeEvents` type

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    );
+  }
+
+  return {
```
use the `Subscription` type instead of casting each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
```diff
@@ -0,0 +1,314 @@
+    return null;
+  }
+
+  return {
```
use the `SubscriptionCycle` type instead of casting each column

habdelra commented on packages/realm-server/billing/billing_queries.ts:
rename this module to `billing-quries.ts` to be consistent (no underscores in module names)

habdelra commented on packages/realm-server/server.ts:
```diff
@@ -656,6 +658,13 @@ export class RealmServer {
     throw new Error(`Can not determine the matrix registration secret`);
   }
 
+  private handleStripeWebhook = async (ctxt: Koa.Context, _next: Koa.Next) => {
```
is there a way to auth the request so not just anyone can call this endpoint? like a list of valid IP's and localhost? or something similar?

habdelra replied:
nvm, i just noticed there is a `STRIPE_WEBHOOK_SECRET` in the handleStripeWebhook

jurgenwerk replied:
Yeah, that will allow us to secure this endpoint. I'll have a PR for adding that ENV var soon

habdelra commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+    stripe_customer_id: stripeCustomerId,
+  });
+  let result = await query(dbAdapter, [
+    `INSERT INTO users (matrix_user_id, stripe_customer_id) VALUES`,
```
use `nameExpressions` here

habdelra commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+    stripe_plan_id: stripePlanId,
+  });
+  let result = await query(dbAdapter, [
+    `INSERT INTO plans (name, monthly_price, credits_included, stripe_plan_id) VALUES`,
```
use `nameExpressions` here

habdelra commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+    ` RETURNING *`,
+  ] as Expression);
+
+  return {
```
use the `User` type instead of casting each column

habdelra commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+    ` RETURNING *`,
+  ] as Expression);
+
+  return {
```
use the `Plans` type instead of casting each column

habdelra commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+  let results = (await query(dbAdapter, [
+    `SELECT * FROM subscriptions WHERE user_id = `,
+    param(userId),
+  ])) as {
```
use the `Subscription` type instead of casting each column

habdelra commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+    param(subscriptionId),
+  ]);
+
+  return results.map((result) => ({
```
use the `SubscriptionCycle` type instead of casting each column

habdelra commented on packages/realm-server/tests/billing-test.ts:
```diff
@@ -0,0 +1,367 @@
+    param(userId),
+  ]);
+
+  return results.map((result) => ({
```
use the `CreditLedger` type instead of casting each column

habdelra reviewed (changes requested):
please cleanup the types and use the `nameExpressions` instead of throwing them away

habdelra commented:
Just a general observation, it looks like we are forming a layer to update the ledger entry in the DB as a realm-server module. What happens when we need to update the ledger based on ai-bot usage. Are we going to consume modules from the realm server in the ai-bot? That seems wrong architecturally as the realm server package was never meant for consumption (like runtime-common).  We’ve defined a type for the ledger entry table in this PR that we’ll probably want to consume in the ai-bot when we start crediting ai usage. Do we want to actually follow the pattern established around the DB access for indexing where we have the DB queries live in common such that it can be consumed by different packages (in that case the host and realm-server)? That way when it comes time to start having the ai-bot run queries we can use common code between the realm server and ai-bot. 

Also this is related to using DATABASE_URL as an env var on the ai-bot server, in that you would NOT want to use that env var, but rather use PG* vars instead.  (I think @backspace is working that issue).

jurgenwerk committed:
Fix Stripe dependency category

jurgenwerk committed:
Update pnpm lock

jurgenwerk committed:
Simplify

jurgenwerk committed:
Refactor

jurgenwerk committed:
billing_queries -> billing-queries

jurgenwerk committed:
Convert all timestamps to bigints

jurgenwerk commented:
@habdelra to summarize this is the current plan we've been discussing regarding the AI bot:

1. Before starting an AI action, in the bot app, sum the user's credits ledger and check if positive
2. After generating the AI response, Insert a job in the queue so that a worker can: 1) issue a query against open router to determine the cost of the AI generation, and create a ledger entry

If we do it that way then I suppose adding ledger entries will be handled by a worker that lives in the realm server which could allow us to simplify the db access pattern in the ai bot to the bare minimum.

🤔 TBH I'm still wondering if it would be simpler to just expose 2 new secure endpoints in the realm server; one for checking if there are any available credits and one for submitting the consumption writing job...

jurgenwerk committed:
Merge branch 'main' into cs-7386-handle-stripe-subscribed-webhook

jurgenwerk commented:
@backspace @lukemelia @habdelra I addressed all your feedback:

- Made sure all timestamps are now integers
- Used billing entity types instead of casting attributes in objects
- Refactored the big handler function into 3 smaller ones so it's easier to reason about

habdelra commented:
@jurgenwerk my comment around timestamps was that for unix times they can actually just be regular ints not big ints

habdelra commented on packages/realm-server/migrations/1729595943510_billing-adjustments.js:
```diff
@@ -0,0 +1,101 @@
+  ];
+  for (let table of tablesWithCreatedAt) {
+    pgm.alterColumn(table, 'created_at', {
+      default: pgm.func('CURRENT_TIMESTAMP'),
```
won't this return a `timestamp` type and not an int? (i would at probably expect some math here to create unix time ints)

jurgenwerk replied:
Yes, it would. You are looking at the "down" migration, which goes back to timestamps

habdelra replied:
ah my bad!

habdelra commented on packages/realm-server/migrations/1729595943510_billing-adjustments.js:
```diff
@@ -0,0 +1,101 @@
+  for (let { table, columns } of timestampColumns) {
+    for (let column of columns) {
+      pgm.alterColumn(table, column, {
+        type: 'timestamp',
```
not int?

jurgenwerk replied:
Related to my other answer to a similar question, so it should be fine.

habdelra commented on packages/realm-server/migrations/1729595943510_billing-adjustments.js:
```diff
@@ -0,0 +1,101 @@
+  ];
+  for (let table of tablesWithCreatedAt) {
+    pgm.alterColumn(table, 'created_at', {
+      default: pgm.func('EXTRACT(epoch FROM CURRENT_TIMESTAMP)::bigint'),
```
does this convert to unix time is is this epoch ms?

jurgenwerk replied:
extract(epoch FROM ...) returns seconds, so it should be alright for unix time.

habdelra commented on packages/realm-server/migrations/1729595943510_billing-adjustments.js:
```diff
@@ -0,0 +1,101 @@
+      pgm.alterColumn(table, column, {
+        type: 'bigint',
+        notNull: column !== 'ended_at',
+        using: 'extract(epoch from ' + column + ')::bigint',
```
is this unix time or ephoc ms?

jurgenwerk replied:
Unix (seconds)

habdelra commented:
> @habdelra to summarize this is the current plan we've been discussing regarding the AI bot:
> 
> 1. Before starting an AI action, in the bot app, sum the user's credits ledger and check if positive
> 2. After generating the AI response, Insert a job in the queue so that a worker can: 1) issue a query against open router to determine the cost of the AI generation, and create a ledger entry
> 
> If we do it that way then I suppose adding ledger entries will be handled by a worker that lives in the realm server which could allow us to simplify the db access pattern in the ai bot to the bare minimum.
> 
> 🤔 TBH I'm still wondering if it would be simpler to just expose 2 new secure endpoints in the realm server; one for checking if there are any available credits and one for submitting the consumption writing job...

ok, so how does the AI bot actually check the sum of the user's credits? direct from the DB or via a webservice?

habdelra commented on packages/realm-server/billing/billing-queries.ts:
```diff
@@ -0,0 +1,327 @@
+    ` VALUES`,
+    ...addExplicitParens(separatedByCommas(valueExpressions)),
+  ] as Expression);
+}
```
in the expressions.ts module we have a function `upsert` that just takes in nameExpressions and valueExpressions so that the calls can be very simple, like:

``` ts   
await this.query(upsert(
        'boxel_index',
        'boxel_index_pkey',
        nameExpressions,
        valueExpressions,
      ),
    );
```

let's make a function for `insert` that also takes in a `nameExpression` and a `valueExpression`, so that we can just write:
```ts
await query(insert('table_name', nameExpressions, valueExpressions))
```
then we could remove tons of boilerplate code in this module

jurgenwerk replied:
Seems like a good idea, I'll do it

jurgenwerk replied:
Done https://github.com/cardstack/boxel/pull/1719/commits/9e78818a27854a49bd0a221dc3c1dfa005014b81

jurgenwerk committed:
Big ints can actually be ints (for unix time in seconds)

jurgenwerk commented:
> @jurgenwerk my comment around timestamps was that for unix times they can actually just be regular ints not big ints

I was worried about the [Year 2038 problem](https://en.wikipedia.org/wiki/Year_2038_problem) 😃

> ok, so how does the AI bot actually check the sum of the user's credits? direct from the DB or via a webservice?

In our planning we defined it should get it from the DB. But I have some second thoughts about this after your brought up the dilemmas regarding AI bot and the DB (shared db configs, migrations,...). If we used http calls to the server to get and report consumption we could perhaps avoid this db-related work in the ai bot?

habdelra commented:
> > @jurgenwerk my comment around timestamps was that for unix times they can actually just be regular ints not big ints
> 
> I was worried about the [Year 2038 problem](https://en.wikipedia.org/wiki/Year_2038_problem) 😃
> 
> > ok, so how does the AI bot actually check the sum of the user's credits? direct from the DB or via a webservice?
> 
> In our planning we defined it should get it from the DB. But I have some second thoughts about this after your brought up the dilemmas regarding AI bot and the DB (shared db configs, migrations,...). If we used http calls to the server to get and report consumption we could perhaps avoid this db-related work in the ai bot?

year 2038 is only a problem for 32 bit architectures tho...

habdelra commented:
> > @jurgenwerk my comment around timestamps was that for unix times they can actually just be regular ints not big ints
> 
> I was worried about the [Year 2038 problem](https://en.wikipedia.org/wiki/Year_2038_problem) 😃
> 
> > ok, so how does the AI bot actually check the sum of the user's credits? direct from the DB or via a webservice?
> 
> In our planning we defined it should get it from the DB. But I have some second thoughts about this after your brought up the dilemmas regarding AI bot and the DB (shared db configs, migrations,...). If we used http calls to the server to get and report consumption we could perhaps avoid this db-related work in the ai bot?

how about we talk about in the boxel meeting? i think the DB is fine, but it does mean being more smart about where the shared code lives.

jurgenwerk committed:
DRY up insert statements

jurgenwerk committed:
Not needed
